### PR TITLE
Fix #5339: cache dtype.__hash__.

### LINKED ISSUE #5339: cache dtype.__hash__
It would be nice to cache the hash value of dtypes, since computation can be non-trivial for complex (e.g. structured) dtypes.

I would like to contribute this but I had some trouble running the test suite, and now I'm not able to build Numpy anymore:

```
$ python3.4 setup.py build
Running from numpy source directory.
Traceback (most recent call last):
  File "setup.py", line 251, in <module>
    setup_package()
  File "setup.py", line 235, in setup_package
    from numpy.distutils.core import setup
  File "/home/antoine/np/numpy/distutils/__init__.py", line 37, in <module>
    from numpy.testing import Tester
  File "/home/antoine/np/numpy/testing/__init__.py", line 14, in <module>
    from .utils import *
  File "/home/antoine/np/numpy/testing/utils.py", line 17, in <module>
    from numpy.core import float32, empty, arange, array_repr, ndarray
  File "/home/antoine/np/numpy/core/__init__.py", line 14, in <module>
    from . import multiarray
ImportError: cannot import name 'multiarray'
```

### PR DESCRIPTION
**Pull request by pitrou**
Fix #5339: cache `dtype.__hash__`

### CODE REVIEW

File: numpy/core/include/numpy/ndarraytypes.h [hunk 1]
```diff
@@ -619,6 +620,10 @@ typedef struct _PyArray_Descr {
          * for NumPy 1.7.0.
          */
         NpyAuxData *c_metadata;
+        /* Cached hash value (-1 if not yet computed).
+         * This was added for NumPy 2.0.0.
+         */
+        npy_hash_t hash;
```
Comment (pitrou) [inline numpy/core/include/numpy/ndarraytypes.h]: I'm less knowledgeable than you guys :) What should be the way forward?

Comment (juliantaylor) [inline numpy/core/include/numpy/ndarraytypes.h]: one way would be to implement the hiding of the object internals, another wait until someone else does it, but that might take a while.
Unfortunately the dtype internals are probably far more commonly used than the ufunc internals I want to hide. This probably gives us much less wiggle room for breaking the ABI.
Unless someone finds a way to do so without breaking it, maybe some kind of magic number in one of the existing fields could work.

Comment (juliantaylor) [inline numpy/core/include/numpy/ndarraytypes.h]: maybe I am also just overly cautious, in light of no better idea if you want to work on it I think the approach of adding a second private struct that embeds the public struct is worthwhile. Its only problem is that we technically can't trust the hash to be correct if others to weird unusual with the public part.

Comment (pitrou) [inline numpy/core/include/numpy/ndarraytypes.h]: Would the general scheme work like PyArray_Fields currently does? Which dtype fields should be private vs. public exactly?

Comment (pitrou) [inline numpy/core/include/numpy/ndarraytypes.h]: Ping on my questions above :)

**Comment by pitrou** [comment]
Re-re-ping.

**Comment by charris** [comment]
@juliantaylor @njsmith Thoughts? I agree with Julian that the descr internals are accessed in a *lot* of code, and that embedding the structure might be a way forward, i.e., dividing it into private and public parts. Long term, we might want to add functions for access, but I don't think we can require them in any time frame less than several/many years.

**Comment by njsmith** [comment]
I'm all for doing some more general rearrangements to how the dtype struct
works, but that's another discussion. I am totally unbothered by adding a
field to the end of the struct now and deferring any more cleanups until
later.

I'd worry more about correctness in face of the fact that dtypes are
mutable in general. (This is horrible but I'm pretty sure it's true that
e.g. you can modify the field names of a structured dtype.) One option
would be to only cache the hashes of known-safe dtypes like the atomic
types. Another would be to tell people sorry, if you go mutating dtypes in
place then you deserve whatever you get.

Making the uninitialized value 0 as discussed above might be a minor
improvement.

**Comment by pitrou** [comment]
Could you show me an example of how to mutate a dtype?

**Comment by pv** [comment]
```
x = np.dtype([('a', 'f8')])
x.names = ('foo',)
```

**Comment by pitrou** [comment]
Thanks. Apparently `names` is the only settable attribute on dtype, so we can invalidate the cached hash there (and in `__setstate__` as well).

**Comment by charris** [comment]
Hmm, is there a good reason for the names being mutable? Should we consider removing that ability?

**Comment by pitrou** [comment]
As a matter of fact, there's the following comment in the setter for the `names` attribute:

```
    /*
     * This deprecation has been temporarily removed for the NumPy 1.7
     * release. It should be re-added after the 1.7 branch is done,
     * and a convenience API to replace the typical use-cases for
     * mutable names should be implemented.
     *
     * if (DEPRECATE("Setting NumPy dtype names is deprecated, the dtype "
     *                "will become immutable in a future version") < 0) {
     *     return -1;
     * }
     */
```

**Comment by pitrou** [comment]
(the deprecation was added in 935c5b0e and commented out in ff4aef41)

**Comment by pitrou** [comment]
I've updated the PR to invalidate the cache when necessary, and add tests.

**Comment by charris** [comment]
Why the merge from master? Few comments

* Should probably uncomment the deprecation and test for it, maybe as a separate PR. I assume it was commented out for a reason,  so someone must be using the feature. Can views be used instead?
*  What was the circular include dependency? Is there a cleaner way to fix it? In any case, the name of the new include file doesn't sound right, it is less than minimal ;)

**Comment by pitrou** [comment]
> What was the circular include dependency?

npy_3kcompat.h includes ndarraytypes.h includes ndarrayobject.h includes npy_3kcompat.h.

> Is there a cleaner way to fix it?

Depends what you call "cleaner" :-) Perhaps move the "npy_hash_t" typedef into npy_common.h?

**Comment by charris** [comment]
Mailing list thread on the topic http://thread.gmane.org/gmane.comp.python.numeric.general/43047/focus=43119. Note the comment that the dtype was made mutable to work around the inablility to do the renames with a *.view. We might want to take a look at that, as it seems that would be the preferred alternative.

**Comment by charris** [comment]
@pitrou Yeah, let's put `npy_hash_t` in `npy_common.h`. It's a late addition to the compat file and that file contains no other typedefs. After that I think you can rebase cleanly on master and squash the commits into one. Please use the `ENH: ` tag on the commit summary line, add a note to the 1.10 release notes mentioning the change in the struct (although I don't think it will cause any problems except for Cython warnings). Please expand the rest of the commit message, saying what you have done and include the benchmark results. Finish the commit message with `Closes #5339`. After all that, do a force push ;)

**Comment by pitrou** [comment]
@charris, ok, I think I have done everything you asked for.

**Comment by charris** [comment]
LGTM, thanks @pitrou. @juliantaylor @njsmith This simply extends the struct at the end. I don't think this will break anything in practice. If we intend to hide some stuff we should settle how to do that soon.

**Comment by juliantaylor** [comment]
I don't like starting the 1.7 cython warning nonsense again and a (minor) abi break for a performance improvement most users are never going to see ...
sorting this out once and for all needs to be a 1.10 blocker

**Comment by pitrou** [comment]
Perhaps Cython simply needs fixing? What is the warning about exactly?

**Comment by juliantaylor** [comment]
change of structure size, aka abi break. its probably too noisy as the abi is only broken for (hopefully) very uncommon cases. Changing cython will not help much as the current structure size is already included in existing binaries which you'd need to recompile to remove it.

**Comment by rgommers** [comment]
Details at https://github.com/numpy/numpy/pull/432. That was/is indeed annoying as hell.

**Comment by njsmith** [comment]
The change in #432 to make cython shut up about this nonsense is still in
effect, right? Doesn't that take care of things?

On Sat, Apr 4, 2015 at 4:21 PM, Ralf Gommers <notifications@github.com>
wrote:

> Details at #432 <https://github.com/numpy/numpy/pull/432>. That was/is
> indeed annoying as hell.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5343#issuecomment-89683190>.
>



-- 
Nathaniel J. Smith -- http://vorpus.org

**Comment by pitrou** [comment]
PyObjects should only be allocated dynamically and passed by reference (the exception is static definition of a PyTypeObject), so I don't know why a size change would be a problem. CPython can change the size of builtin objects from one version to another: does Cython complain about that?

**Comment by njsmith** [comment]
It's never made a huge amount of sense to me either. You might bring it up
with the Cython devs...

(Does CPython really change the size of builtin objects in point releases?
IIRC Cython extensions don't restrict themselves to the stable ABI, so they
have to be recompiled for major releases anyway.)

On Sat, Apr 4, 2015 at 4:26 PM, Antoine Pitrou <notifications@github.com>
wrote:

> PyObjects should only be allocated dynamically and passed by reference
> (the exception is static definition of a PyTypeObject), so I don't know why
> a size change would be a problem. CPython can change the size of builtin
> objects from one version to another: does Cython complain about that?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/5343#issuecomment-89685346>.
>



-- 
Nathaniel J. Smith -- http://vorpus.org

**Comment by rgommers** [comment]
@njsmith I just checked, the filters in ``numpy/__init__.py`` are still there and the string it filters on is still unchanged. So should be OK for this PR. There's no filter for numpy.ndarray though, maybe we should add that proactively just in case.

**Comment by pitrou** [comment]
> Does CPython really change the size of builtin objects in point releases?

No, normally not. Only in feature releases.

**Comment by njsmith** [comment]
> There's no filter for numpy.ndarray though, maybe we should add that proactively just in case.

[I agree](https://github.com/numpy/numpy/pull/432#issuecomment-8397439) :-)

**Comment by rgommers** [comment]
Okay, done in gh-5750.

**Pull request merged**